libs/application/templates/directorate-of-immigration/citizenship/src/fields/ResidenceCountries/index.tsx

```diff
@@ -0,0 +1,222 @@
+    application.externalData,
+    'currentCountryOfResidenceList.data',
+    [],
+  ) as CountryOfResidenceViewModel[]
+
+  const [hasLivedAbroad, setHasLivedAbroad] = useState<string>(
+    getValueViaPath(
+      application.answers,
+      'countriesOfResidence.hasLivedAbroad',
+      preRegisteredCountries.length > 0 ? YES : '',
+    ) as string,
+  )
+
+  const [selectedCountries, setSelectedCountries] = useState<
+    ExtendedCountryProps[]
+  >(
+    getValueViaPath(
+      application.answers,
+      'countriesOfResidence.selectedAbroadCountries',
+      [],
+    ) as ExtendedCountryProps[],
+  )
+
+  const [filteredSelectedCountries, setFilteredSelectedCountries] = useState<
+    Array<ExtendedCountryProps>
+  >([])
+
+  useEffect(() => {
+    const notDuplicateAnswers = selectedCountries.filter(
+      (x) =>
+        preRegisteredCountries.findIndex(
+          (y) => y.countryId === parseInt(x.countryId),
+        ) === -1,
+    )
+
+    const combinedLists = [
+      ...preRegisteredCountries
+        .filter((x) => !!x.countryId)
+        .map((i) => {
+          return {
+            countryId: i.countryId?.toString(),
+            wasRemoved: 'false',
+            readOnly: true,
+          } as ExtendedCountryProps
+        }),
+      ...notDuplicateAnswers,
+    ]
+    setSelectedCountries(combinedLists)
+  }, [])
+
+  useEffect(() => {
+    setFilteredSelectedCountries(
+      selectedCountries.filter((x) => x.wasRemoved !== 'true'),
+    )
+  }, [selectedCountries])
+
+  const handleAdd = () =>
+    setSelectedCountries([
+      ...selectedCountries,
+      {
+        countryId: '',
+        wasRemoved: 'false',
+      },
+    ])
+
+  const handleRemoveAll = () => {
+    setSelectedCountries(
+      selectedCountries.map((x) => {
+        return { ...x, wasRemoved: 'true' }
+      }),
+    )
+  }
+
+  const handleRemove = (pos: number) => {
+    if (pos > -1) {
+      setSelectedCountries(
+        selectedCountries.map((country, index) => {
+          if (index === pos) {
+            return { ...country, wasRemoved: 'true' }
+          }
+          return country
+        }),
+      )
+    }
+  }
+
+  const addCountryToList = (newCountry: string, newIndex: number) => {
+    setSelectedCountries(
+      selectedCountries.map((country, index) => {
+        if (newIndex === index) {
+          return {
+            ...country,
+            countryId: newCountry,
+          }
+        }
+        return country
+      }),
+    )
+  }
+
+  const handleLiveAbroadChange = (value: string) => {
+    setHasLivedAbroad(value)
+
+    if (value === NO) {
+      handleRemoveAll()
+    } else if (preRegisteredCountries.length === 0) {
+      handleAdd()
+    }
+  }
+
+  return (
+    <Box>
+      <DescriptionText
+        text={information.labels.countriesOfResidence.questionTitle}
+        textProps={{
+          as: 'h5',
+          fontWeight: 'semiBold',
+          marginBottom: 3,
+        }}
+      />
+      <RadioController
```

Comment:
No because I have an onSelect on the radioController that is triggering a function and changing some data. I can't do that if I have it outside of the custom component.. right?

Reply:
Yes you can monitor values from outside the component by using react-hook-form watch. You can also set the values by using their setValue function.

Here's the diff showing the exact changes on how this would kinda work, there is a bug where if you go back to the screen an additional country field gets added but ive got to go and dont want to slow you all down 😄 

```
ResidenceCountries/index.tsx
import { FC, useEffect, useState } from 'react'
import { FieldBaseProps } from '@island.is/application/types'
import { Box, Button } from '@island.is/island-ui/core'
import { ResidenceCountriesRepeaterItem } from './ResidenceCountriesRepeaterItem'
import { CountryOfResidence } from '../../shared'
import { NO, YES } from '@island.is/application/core'
import { information } from '../../lib/messages'
import { useLocale } from '@island.is/localization'
import { useFormContext } from 'react-hook-form'

export const ResidenceCountries: FC<FieldBaseProps> = (props) => {
  const { watch, setValue } = useFormContext()
  const { formatMessage } = useLocale()
  const hasLivedAbroad = watch('countriesOfResidenceAnswer')
  const selectedCountries = watch(
    'countriesOfResidence.selectedAbroadCountries',
  ) as CountryOfResidence[]

  const [filteredSelectedCountries, setFilteredSelectedCountries] = useState(
    selectedCountries.filter((x) => x.wasRemoved !== 'true'),
  )

  const setSelectedCountries = (
    filteredSelectedCountries: CountryOfResidence[],
  ) => {
    setValue(
      'countriesOfResidence.selectedAbroadCountries',
      filteredSelectedCountries,
    )
  }

  useEffect(() => {
    setFilteredSelectedCountries(
      selectedCountries.filter((x) => x.wasRemoved !== 'true'),
    )
  }, [selectedCountries])

  useEffect(() => {
    handleLiveAbroadChange()
  }, [hasLivedAbroad])

  const handleAdd = () =>
    setSelectedCountries([
      ...selectedCountries,
      {
        countryId: '',
        wasRemoved: 'false',
      },
    ])

  const handleRemoveAll = () => {
    setSelectedCountries(
      selectedCountries.map((x) => {
        return { ...x, wasRemoved: 'true' }
      }),
    )
  }

  const handleRemove = (pos: number) => {
    if (pos > -1) {
      setSelectedCountries(
        selectedCountries.map((country, index) => {
          if (index === pos) {
            return { ...country, wasRemoved: 'true' }
          }
          return country
        }),
      )
    }
  }

  const addCountryToList = (newCountry: string, newIndex: number) => {
    setSelectedCountries(
      selectedCountries.map((country, index) => {
        if (newIndex === index) {
          return {
            ...country,
            countryId: newCountry,
          }
        }
        return country
      }),
    )
  }

  const handleLiveAbroadChange = () => {
    if (hasLivedAbroad === NO) {
      handleRemoveAll()
    } else {
      handleAdd()
    }
  }

  return (
    <Box>
      <Box>
        {selectedCountries.map((field, index) => {
          const position = filteredSelectedCountries.indexOf(field)
          return (
            <ResidenceCountriesRepeaterItem
              id={`${props.field.id}.selectedAbroadCountries`}
              repeaterField={field}
              index={index}
              key={`countrySelect-${index}`}
              handleRemove={handleRemove}
              itemNumber={position}
              addCountryToList={addCountryToList}
              {...props}
            />
          )
        })}
        {hasLivedAbroad === YES && (
          <Box paddingTop={4}>
            <Button
              variant="ghost"
              icon="add"
              iconType="outline"
              fluid
              size="large"
              onClick={handleAdd}
            >
              {formatMessage(
                information.labels.countriesOfResidence.buttonTitle,
              )}
            </Button>
          </Box>
        )}
      </Box>
    </Box>
  )
}
```


CountriesOfResidenceSubSection.ts
```
import {
  buildMultiField,
  buildSubSection,
  buildCustomField,
  buildDescriptionField,
  buildRadioField,
} from '@island.is/application/core'
import { information } from '../../../lib/messages'
import { Routes } from '../../../lib/constants'

export const CountriesOfResidenceSubSection = buildSubSection({
  id: Routes.COUNTRIESOFRESIDENCE,
  title: information.labels.countriesOfResidence.subSectionTitle,
  children: [
    buildMultiField({
      id: Routes.COUNTRIESOFRESIDENCE,
      title: information.labels.countriesOfResidence.pageTitle,
      children: [
        buildDescriptionField({
          id: 'countriesOfResidenceDescription',
          title: information.labels.countriesOfResidence.questionTitle,
          titleVariant: 'h5',
        }),
        buildRadioField({
          id: 'countriesOfResidenceAnswer',
          title: '',
          largeButtons: true,
          options: [
            {
              value: 'yes',
              label: information.labels.radioButtons.radioOptionYes,
            },
            {
              value: 'no',
              label: information.labels.radioButtons.radioOptionNo,
            },
          ],
        }),
        buildDescriptionField({
          id: 'countriesOfResidenceDescription2',
          title: information.labels.countriesOfResidence.countryListTitle,
          titleVariant: 'h5',
          condition: (formValue) => {
            return formValue.countriesOfResidenceAnswer === 'yes'
          },
        }),
        buildCustomField({
          id: 'countriesOfResidence',
          title: '',
          component: 'ResidenceCountries',
          condition: (formValue) => {
            return formValue.countriesOfResidenceAnswer === 'yes'
          },
        }),
      ],
    }),
  ],
})
```

Reply:
one problem here, I need the disabled option on the radioField to depend on data from the externalData
 buildRadioField({
          id: 'countriesOfResidence.hasLivedAbroad',
          title: '',
          width: 'half',
          defaultValue: (application: Application) => {
            const preRegisteredCountries = getValueViaPath(
              application.externalData,
              'currentCountryOfResidenceList.data',
              [],
            ) as CountryOfResidenceViewModel[]
            return preRegisteredCountries.length > 0
          },
          disabled: (application: Application) => {
            const preRegisteredCountries = getValueViaPath(
              application.externalData,
              'currentCountryOfResidenceList.data',
              [],
            ) as CountryOfResidenceViewModel[]
            return preRegisteredCountries.length > 0
          },
          options: [
            {
              value: YES,
              label: information.labels.radioButtons.radioOptionYes,
            },
            {
              value: NO,
              label: information.labels.radioButtons.radioOptionNo,
            },
          ],
        }),
        
       Like this, but the disabled field only takes boolean | undefined
       I don't know a work around this..